pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1542 @@
+	servicePortEndpointChainNamePrefix      = "endpoint-"
+	servicePortEndpointAffinityNamePrefix   = "affinity-"
+)
+
+// servicePortChainNameBase returns the base name for a chain for the given ServicePort.
+func servicePortChainNameBase(servicePortName *proxy.ServicePortName, protocol string) string {
+	// nftables chains can contain the characters [A-Za-z0-9_./-]
+
+	// Namespace/Service/Port names are all, roughly /[a-z][a-z-]{0,62}/
+	// Protocol is /(tcp|udp|sctp)/
+	// So we can use everything unmodified.
+
+	name := fmt.Sprintf("%s/%s/%s/%s",
+		servicePortName.NamespacedName.Namespace,
+		servicePortName.NamespacedName.Name,
+		protocol,
+		servicePortName.Port,
+	)
+
+	// Prefix a hash of the full name so that (a) the name is still unique if we have
+	// to truncate the end, and (b) it's easier to visually distinguish chains that
+	// would otherwise have nearly identical names (e.g., different endpoint chains
+	// for a given service that differ in only a single digit).
+	hash := sha256.Sum256([]byte(name))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	if len(name) > chainNameBaseLengthMax-10 {
+		name = name[:chainNameBaseLengthMax-13] + "..."
+	}
+	return encoded[:8] + "-" + name
+}
+
+// servicePortEndpointChainNameBase returns the suffix for chain names for the given endpoint
+func servicePortEndpointChainNameBase(servicePortName *proxy.ServicePortName, protocol, endpoint string) string {
+	// As above in servicePortChainNameBase, Namespace, Service, Port, Protocol, and
+	// EndpointPort are all safe to copy into the chain name directly. But if
+	// EndpointIP is IPv6 then it will contain colons, which aren't allowed in a chain
+	// name. IPv6 IPs are also quite long, but we can't safely truncate them (e.g. to
+	// only the final segment) because (especially for manually-created external
+	// endpoints), we can't know for sure that any part of them is redundant.
+
+	endpointIP, endpointPort, _ := net.SplitHostPort(endpoint)
+	if strings.Contains(endpointIP, ":") {
+		endpointIP = strings.ReplaceAll(endpointIP, ":", ".")
+	}
+
+	name := fmt.Sprintf("%s/%s/%s/%s__%s/%s",
+		servicePortName.NamespacedName.Namespace,
+		servicePortName.NamespacedName.Name,
+		protocol,
+		servicePortName.Port,
+		endpointIP,
+		endpointPort,
+	)
+
+	// Prefix a hash of the full name so that (a) the name is still unique if we have
+	// to truncate the end, and (b) it's easier to visually distinguish chains that
+	// would otherwise have nearly identical names (e.g., different endpoint chains
+	// for a given service that differ in only a single digit).
+	hash := sha256.Sum256([]byte(name))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	if len(name) > chainNameBaseLengthMax-10 {
+		name = name[:chainNameBaseLengthMax-13] + "..."
+	}
+	return encoded[:8] + "-" + name
+}
+
+func isServiceChainName(chainString string) bool {
+	// The chains returned from servicePortChainNameBase and
```

Comment:
are we not prefixing them? `	servicePortPolicyClusterChainNamePrefix = "service-"`

Reply:
We are, but in the middle parts of the PR, there is a chain called `external-services` (corresponding to iptables `KUBE-EXTERNAL-SERVICES`) which would get mistakenly recognized as a service chain because it starts with `external-` (which also corresponds to iptables `KUBE-EXT-`). (The `external-services` chain goes away in a later commit, so I _could_ check prefixes instead after that... Or I could make `external-services` have a different name to avoid the conflict in the first place. But then again, scanning the chain name for a single character seems probably faster than comparing against an array of prefixes?)

Note that the fancy chain names are something I just implemented last week after Tim suggested it in the KEP, so this code is pretty new and maybe less-well-thought-out than some other parts...